rust/datafusion/src/physical_plan/limit.rs

```diff
@@ -200,30 +200,39 @@ pub fn truncate_batch(batch: &RecordBatch, n: usize) -> RecordBatch {
 
 /// A Limit stream limits the stream to up to `limit` rows.
 struct LimitStream {
+    /// The maximum number of rows to produce
     limit: usize,
-    input: SendableRecordBatchStream,
-    // the current count
+    /// The input to read from. This is set to None once the limit is
+    /// reached to enable early termination
+    input: Option<SendableRecordBatchStream>,
+    /// Copy of the input schema
+    schema: SchemaRef,
+    // the current number of rows which have been produced
     current_len: usize,
 }
 
 impl LimitStream {
     fn new(input: SendableRecordBatchStream, limit: usize) -> Self {
+        let schema = input.schema();
```

Comment:
The actual code change for this PR is very small -- the rest of the changes are related to writing a proper test for it

Reply:
I have a vague memory that [FusedStream](https://docs.rs/futures/0.3.13/futures/stream/trait.FusedStream.html) may have something to do with this property (although /noideadog)

Reply:
This is a good point.

One benefit of the this PR over `fuse()` is that this PR will actually drop the input stream (freeing resources) in addition to not calling the input stream again: https://docs.rs/futures-util/0.3.13/src/futures_util/stream/stream/fuse.rs.html#10